### PR TITLE
DiscIO: Add a Volume::ReadSwappedAndShifted function

### PR DESCRIPTION
**Pull request by JosJuice**
This is a fairly common operation, so let's make a utility function for it to cut down on code duplication.

### CODE REVIEW

File: Source/Core/DiscIO/Volume.h [hunk 1]
```diff
@@ -51,6 +51,12 @@ class Volume
       return {};
     return Common::FromBigEndian(temp);
   }
+  std::optional<u64> ReadSwappedAndShifted(u64 offset, const Partition& partition) const
+  {
+    const std::optional<u32> temp = ReadSwapped<u32>(offset, partition);
+    return temp ? static_cast<u64>(*temp) << GetOffsetShift() : std::optional<u64>();
```
Comment (lioncash) [inline Source/Core/DiscIO/Volume.h]: You can use `std::nullopt` instead of `std::optional<u64>()` here as a shorthand.

Comment (JosJuice) [inline Source/Core/DiscIO/Volume.h]: It doesn't seem to work when used inside a ternary operator like this. I get the following error (in VS2017):

```
error C2446: ':': no conversion from 'const std::nullopt_t' to 'u64'
```

**Comment by lioncash** [comment]
It's because the `static_cast<u64>(*temp) << GetOffsetShift()` expression
doesn't result in a std::optional , but a u64. If types are different in
the last two operand placements of a ternary, then implicit conversions
occur, which essentially attempt to deduce a common type between the second
and third operand as the result type (and std::nullopt isn't implicitly
convertible to u64).

**Comment by leoetlino** [comment]
I guess it would work if the second operand were a `std::optional` instead of `u64`. Too bad there's no syntax with less visual noise :/

**Pull request merged**